### PR TITLE
Added subst lemmas, graph coloring example

### PR DESCRIPTION
**Pull request by ccodel**
A bunch of lemmas were added to `Model/Subst.lean`, and the example file from the Formal Cookies talk was added to the `Examples/` folder.

Notes: compare `substL` (line 151) to `subst` (line 185), and see the TODO on line 219.

Better notation/syntax? See lines 49-53.

### CODE REVIEW

File: Examples/GraphColoring.lean [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+  [propfun|
+    ({blue   u}  ∨
+     {red    u}  ∨
+     {green  u}  ∨
+     {yellow u})
+  ]
+
+-- Alternative definition treating CNF clauses as boolean algebras
+def eachVertexGetsAColor' (v : Fin n) : PropFun (ColorVars n) :=
+  (blue v) ⊔ (red v) ⊔ (green v) ⊔ (yellow v)
```
Comment (JamesGallicchio) [inline Examples/GraphColoring.lean:87]: I think I prefer this notation. Maybe we should standardize around this notation? (and remove the other from this example)

We could even add it as syntax for PropForm, even though PropForm is not a booleanalgebra

Comment (JamesGallicchio) [reply Examples/GraphColoring.lean:87]: created #17

File: Examples/GraphColoring.lean [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+
+def adjacentVertexesGetDifferentColors (u v : Fin n) : PropFun (ColorVars n) :=
+  [propfun|
+    (¬ ({blue    u} ∧ {blue   v})) ∧
+    (¬ ({red     u} ∧ {red    v})) ∧
+    (¬ ({green   u} ∧ {green  v})) ∧
+    (¬ ({yellow  u} ∧ {yellow v}))
+  ]
+
+def adjacentVertexesGetDifferentColors' (u v : Fin n) : PropFun (ColorVars n) :=
```
Comment (JamesGallicchio) [inline Examples/GraphColoring.lean:101]: although here i prefer the \not notation. boolean algebra complement is so bad :sob:

File: Examples/GraphColoring.lean [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+def coloringAssignment (C : Coloring n 4) : PropAssignment (ColorVars n) :=
+  fun
+  | .blue   v => C v = ⟨0, by decide⟩
+  | .red    v => C v = ⟨1, by decide⟩
+  | .green  v => C v = ⟨2, by decide⟩
+  | .yellow v => C v = ⟨3, by decide⟩
+
+example (f : ℕ → Prop) (p : Fin 3) (h0 : f 0) (h1 : f 1) (h2 : f 2) : f p.val := by
+  fin_cases p; simp
+  all_goals assumption
```
Comment (JamesGallicchio) [inline Examples/GraphColoring.lean:176]: delete

File: Examples/GraphColoring.lean [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+
+theorem splitFin4 (n : Fin 4) :
```
Comment (JamesGallicchio) [inline Examples/GraphColoring.lean:178]: `fin_cases n` should work directly in context

File: Examples/GraphColoring.lean [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+def K8_CNF : VCnf 8 (graphColoring K8) :=
+  graphColoringCNF K8
+
+def PrintTheCNF : IO Unit :=
+  let cnf := K8_CNF.val.toICnf
+  Solver.Dimacs.printFormula IO.print cnf
+
+-- #eval PrintTheCNF
+
+-- We trust the SAT solver and axiomatize the UNSAT result
```
Comment (JamesGallicchio) [inline Examples/GraphColoring.lean:222]: let's leave a TODO here for when we have a macro that handles introducing this axiom

File: LeanSAT/Model/Subst.lean [hunk 6]
```diff
@@ -38,27 +46,137 @@ def PropForm.subst (p : PropForm ν₁) (f : ν₁ → PropForm ν₂) : PropFor
...
+    : τ ⊨ φ.subst f ↔ τ.subst (⟦f ·⟧) ⊨ φ := by
+  induction φ <;> simp [subst, PropAssignment.subst, *]; rw [PropFun.satisfies_mk]
+
+-- CC: Better to use equivalence or ⟦ ⟧ = ⟦ ⟧?
+theorem subst_congr {φ₁ φ₂} (hφ : equivalent φ₁ φ₂)
+    : ∀ (σ : ν₁ → PropForm ν₂), (⟦φ₁.subst σ⟧ : PropFun _) = ⟦φ₂.subst σ⟧ := by
+  intro σ
+  apply PropFun.ext
+  intro τ
+  -- CC: simp_rw doesn't work here, for some reason...
```
Comment (JamesGallicchio) [inline LeanSAT/Model/Subst.lean:77]: I was running into the same issue, I believe it was an issue with re-instantiated typeclass arguments, which was changed in 4.7 so maybe we should update and see if it fixes that

File: LeanSAT/Model/Subst.lean [hunk 7]
```diff
@@ -38,27 +46,137 @@ def PropForm.subst (p : PropForm ν₁) (f : ν₁ → PropForm ν₂) : PropFor
...
+end substOne /- section -/
+
+end PropForm
+
+/-! # PropFun -/
+
+namespace PropFun
+
+-- CC: A computable, lifting version of subst. Note the mis-ordering of arguments.
+def substL (f : ν₁ → PropForm ν₂) : PropFun ν₁ → PropFun ν₂ :=
```
Comment (JamesGallicchio) [inline LeanSAT/Model/Subst.lean:145]: the diff between `subst` and `substL` (aside from computability) is the type of `f`. maybe add a doc comment explaining this?

File: LeanSAT/Model/Subst.lean [hunk 8]
```diff
@@ -88,6 +209,26 @@ theorem PropFun.satisfies_subst [DecidableEq ν₁]
   simp at this
   exact this.symm
 
+-- CC: Unsure how to prove for fancy `subst`.
+--@[simp] theorem subst_distrib : subst ⟦v⟧ f = f v := by
```
Comment (JamesGallicchio) [inline LeanSAT/Model/Subst.lean:213]: `by  ext; simp [PropAssignment.subst]` :D

**Review by JamesGallicchio — CHANGES_REQUESTED** [review]

File: Examples/GraphColoring.lean [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+    | ⟨n + 4, hn⟩ => by contradiction
+  left_inv := sorry
+  right_inv := sorry
+}
+
+-- Open the namespace so we can use ".blue" instead of "Coloring.blue", etc.
+open ColorVars
+
+-- We sometimes need to provide a lexicographic ordering for our variables
+instance : LinearOrder (ColorVars n) :=
```
[the remark this answers is not in the document]
Comment (ccodel) [reply Examples/GraphColoring.lean:66]: Dunno, but it was included in whatever example I was consulting (probably Pigeon.lean). I'll see what happens if it gets deleted.

File: LeanSAT/Model/Subst.lean (hunk 8 above)
Comment (ccodel) [reply LeanSAT/Model/Subst.lean:213]: We love short proofs that we miss initially.

File: Examples/GraphColoring.lean (hunk 4 above)
Comment (ccodel) [reply Examples/GraphColoring.lean:178]: You would expect it to. However, calling `fin_cases` on the output of the coloring function encounters a Lean error `unexpected syntax: fin_cases (C v)`. Putting a `let Cv := C v` beforehand gives a slightly different error:
```
tactic 'cases' failed, nested error:
dependent elimination failed, failed to solve equation
  { val := 0, isLt := (_ : 0 < 4) } = Cv
```
In my attempts to get a proof ready for formal cookies, externalizing the `fin_cases` in this theorem was the only thing that worked. I welcome additional changes in the future.

File: LeanSAT/Model/Subst.lean [hunk 10]
```diff
@@ -38,27 +46,143 @@ def PropForm.subst (p : PropForm ν₁) (f : ν₁ → PropForm ν₂) : PropFor
...
 
 @[simp]
-theorem PropForm.satisfies_subst [DecidableEq ν] (φ : PropForm ν) (f : ν → PropForm ν') {τ : PropAssignment ν'}
-  : τ ⊨ φ.subst f ↔ τ.subst (⟦f ·⟧) ⊨ φ
-  := by induction φ <;> simp [subst, PropAssignment.subst, *]; rw [PropFun.satisfies_mk]
+theorem satisfies_subst {φ : PropForm ν₁} {f} {τ : PropAssignment ν₂}
+    : τ ⊨ φ.subst f ↔ τ.subst (⟦f ·⟧) ⊨ φ := by
+  induction φ <;> simp [subst, PropAssignment.subst, *]; rw [PropFun.satisfies_mk]
+
+-- CC: Better to use equivalence or ⟦ ⟧ = ⟦ ⟧?
```
[the remark this answers is not in the document]
Comment (ccodel) [reply LeanSAT/Model/Subst.lean]: I've now made the PR consistent by using [.] = [.].

File: Examples/GraphColoring.lean [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+import LeanSAT.Solver.Dimacs
+import LeanSAT.Data.ICnf
+
+import Mathlib.Data.Fintype.Prod
+import Mathlib.Data.Prod.Lex
+import Mathlib.Data.Sum.Order
+import Mathlib.Combinatorics.SimpleGraph.Basic
+import Mathlib.Tactic
+
+namespace Array
```
[the remark this answers is not in the document]
Comment (ccodel) [reply Examples/GraphColoring.lean]: `List.finRange` is a Mathlib function, so moved to `ToMathlib`. But yes, we can address this later to get a more efficient version.

File: Examples/GraphColoring.lean [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+    | ⟨0, _⟩ => .blue v
+    | ⟨1, _⟩ => .red v
+    | ⟨2, _⟩ => .green v
+    | ⟨3, _⟩ => .yellow v
+    | ⟨n + 4, hn⟩ => by contradiction
+  left_inv := sorry
+  right_inv := sorry
+}
+
+-- Open the namespace so we can use ".blue" instead of "Coloring.blue", etc.
```
[the remark this answers is not in the document]
Comment (ccodel) [reply Examples/GraphColoring.lean]: Comment changed. No good comment goes unpunished.

File: Examples/GraphColoring.lean (hunk 9 above)
[the remark this answers is not in the document]
Comment (ccodel) [reply Examples/GraphColoring.lean]: Nothing happened when deleted. So it's been deleted.

File: Examples/GraphColoring.lean (hunk 5 above)
Comment (ccodel) [reply Examples/GraphColoring.lean:202]: New comment left.

File: LeanSAT/Model/Subst.lean (hunk 7 above)
Comment (ccodel) [reply LeanSAT/Model/Subst.lean]: Left a doc comment, and matched the argument ordering to the other `subst` functions.

File: LeanSAT/Model/Subst.lean (hunk 8 above)
Comment (ccodel) [reply LeanSAT/Model/Subst.lean]: Proof added.

File: LeanSAT/Model/Subst.lean (hunk 6 above)
Comment (ccodel) [reply LeanSAT/Model/Subst.lean]: Comment left to address the `simp_rw` bug when we upgrade Lean version.

File: Examples/GraphColoring.lean (hunk 2 above)
Comment (ccodel) [reply Examples/GraphColoring.lean:81]: I agree. We can adjust syntax later.

File: Examples/GraphColoring.lean (hunk 3 above)
Comment (ccodel) [reply Examples/GraphColoring.lean]: Deleted.

**Review by JamesGallicchio — APPROVED** [review]

**Pull request merged**